File: src/CalcManager/CalculatorHistory.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include <cassert>
#include "CalculatorHistory.h"

using namespace std;
using namespace CalculationManager;

namespace
{
    static wstring GetGeneratedExpression(const vector<pair<wstring, int>>& tokens)
    {
        wstring expression;
        bool isFirst = true;

        for (auto const& token : tokens)
        {
            if (isFirst)
            {
                isFirst = false;
            }
            else
            {
                expression += L' ';
            }
            expression.append(token.first);
        }

        return expression;
    }
}

CalculatorHistory::CalculatorHistory(size_t maxSize)
    : m_maxHistorySize(maxSize)
{
}

unsigned int CalculatorHistory::AddToHistory(
    _In_ shared_ptr<vector<pair<wstring, int>>> const& tokens,
    _In_ shared_ptr<vector<shared_ptr<IExpressionCommand>>> const& commands,
    wstring_view result)
{
    shared_ptr<HISTORYITEM> spHistoryItem = make_shared<HISTORYITEM>();

    spHistoryItem->historyItemVector.spTokens = tokens;
    spHistoryItem->historyItemVector.spCommands = commands;
    spHistoryItem->historyItemVector.expression = GetGeneratedExpression(*tokens);
    spHistoryItem->historyItemVector.result = wstring(result);
    return AddItem(spHistoryItem);
}

unsigned int CalculatorHistory::AddItem(_In_ shared_ptr<HISTORYITEM> const& spHistoryItem)
{
    if (m_historyItems.size() >= m_maxHistorySize)
    {
        m_historyItems.erase(m_historyItems.begin());
    }

    m_historyItems.push_back(spHistoryItem);
    return static_cast<unsigned>(m_historyItems.size() - 1);
}

bool CalculatorHistory::RemoveItem(unsigned int uIdx)
{
    if (uIdx < m_historyItems.size())
    {
        m_historyItems.erase(m_historyItems.begin() + uIdx);
        return true;
    }

    return false;
}

vector<shared_ptr<HISTORYITEM>> const& CalculatorHistory::GetHistory()
{
    return m_historyItems;
}

shared_ptr<HISTORYITEM> const& CalculatorHistory::GetHistoryItem(unsigned int uIdx)
{
    assert(uIdx < m_historyItems.size());
    return m_historyItems.at(uIdx);
}

void CalculatorHistory::ClearHistory()
{
    m_historyItems.clear();
}

File: src/CalcManager/CalculatorHistory.rs
use std::collections::VecDeque;
use std::sync::Arc;

pub struct HistoryItemVector {
    pub tokens: Arc<Vec<(String, i32)>>,
    pub commands: Arc<Vec<Arc<dyn IExpressionCommand>>>,
    pub expression: String,
    pub result: String,
}

pub struct HistoryItem {
    pub history_item_vector: HistoryItemVector,
}

pub struct CalculatorHistory {
    max_history_size: usize,
    history_items: VecDeque<Arc<HistoryItem>>,
}

impl CalculatorHistory {
    pub fn new(max_size: usize) -> Self {
        Self {
            max_history_size: max_size,
            history_items: VecDeque::new(),
        }
    }

    pub fn add_to_history(
        &mut self,
        tokens: Arc<Vec<(String, i32)>>,
        commands: Arc<Vec<Arc<dyn IExpressionCommand>>>,
        result: &str,
    ) -> u32 {
        let history_item = Arc::new(HistoryItem {
            history_item_vector: HistoryItemVector {
                tokens: tokens.clone(),
                commands: commands.clone(),
                expression: Self::get_generated_expression(&tokens),
                result: result.to_string(),
            },
        });
        self.add_item(history_item)
    }

    pub fn add_item(&mut self, history_item: Arc<HistoryItem>) -> u32 {
        if self.history_items.len() >= self.max_history_size {
            self.history_items.pop_front();
        }
        self.history_items.push_back(history_item);
        (self.history_items.len() - 1) as u32
    }

    pub fn remove_item(&mut self, index: u32) -> bool {
        if (index as usize) < self.history_items.len() {
            self.history_items.remove(index as usize);
            true
        } else {
            false
        }
    }

    pub fn get_history(&self) -> &VecDeque<Arc<HistoryItem>> {
        &self.history_items
    }

    pub fn get_history_item(&self, index: u32) -> Option<&Arc<HistoryItem>> {
        self.history_items.get(index as usize)
    }

    pub fn clear_history(&mut self) {
        self.history_items.clear();
    }

    fn get_generated_expression(tokens: &[(String, i32)]) -> String {
        tokens
            .iter()
            .map(|(token, _)| token.clone())
            .collect::<Vec<_>>()
            .join(" ")
    }
}
